sql-plugin/src/main/330db/scala/org/apache/spark/sql/rapids/execution/GpuExecutorBroadcastHelper.scala

```diff
@@ -54,23 +67,41 @@ object GpuExecutorBroadcastHelper extends Arm {
    * Given an RDD of ColumnarBatch containing broadcast data from a shuffle, get the 
    * fully coalesced ColumnarBatch loaded on to the GPU.
    *
-   * @param buildRelation - the executor broadcast produced by a shuffle exchange
+   * This method uses getSingleBatchWithVerification to handle the empty relation case 
+   * and the unexpected scenario of receiving more than a single batch
+   *
+   * @param shuffleData - the shuffle data from the executor for the "executor broadcast"
    * @param buildSchema - the schema expected for the output
    * @param buildOutput - the output attributes expected in case we receieve an empty RDD
    * @param metricsMap - metrics to populate while the shuffle coalesce iterator loads 
    *                     data from I/O to the device
    * @param targetSize - the target batch size in bytes for loading data on to the GPU 
    */
   def getExecutorBroadcastBatch(
-      buildRelation: RDD[ColumnarBatch], 
+      shuffleData: RDD[ColumnarBatch], 
       buildSchema: StructType,
       buildOutput: Seq[Attribute],
       metricsMap: Map[String, GpuMetric],
       targetSize: Long): ColumnarBatch = {
 
-    val it = shuffleCoalesceIterator(buildRelation, buildSchema, metricsMap, targetSize)
+    val it = shuffleCoalesceIterator(shuffleData, buildSchema, metricsMap, targetSize)
     // In a broadcast hash join, right now the GPU expects a single batch
     ConcatAndConsumeAll.getSingleBatchWithVerification(it, buildOutput)
   }
 
+  /**
+   * Given an RDD of ColumnarBatch containing broadcast data from a shuffle, get the 
+   * number of rows that the received batch contains
+   */
+  def getExecutorBroadcastBatchNumRows(shuffleData: RDD[ColumnarBatch]): Int = {
```

Comment:
Why are we computing this from the `RDD[ColumnarBatch]` instead of reading all of the data into a single table, saving it, and then computing the number of rows? The broadcast equivalent has read the data in and cached it. I am fairly sure that this is going to read the shuffle data twice. The problem with reading the shuffle data twice on a BNLJ is that this is the join of last resort.  Normal broadcasts are likely to be just a few MiB of data. Doing the shuffle twice is not the end of the world. Still not perfect but not a deal breaker.  Here this join gets added in when no other join will work. I don't think we are taking the shuffled exchange in those cases, but it just feels like a waste to do it twice.